commands/docker/command.go

```diff
@@ -0,0 +1,137 @@
+		// repo, we must enforce that we link all resolved projects.
+		// Doing it inline as it shouldn't be called by any other
+		// part of the application bare this part.
+		"node": func(project string) ([]string, error) {
+			node.LinkLocalDeps()
+			return node.Resolve(project)
+		},
+	}
+}
+
+// Config contains all the required information
+// in order to build the given application as a
+// docker image
+type Config struct {
+	// Application is always assumed to be the base name of the current directory
+	Application string
+	// Type allows for correct resolution of required resources
+	Type      string   `yaml:"Type"`
+	BuildArgs []string `yaml:"BuildArgs"`
+	Resources []string `yaml:"Resources"`
+	Template  []byte
+}
+
+// CreateConfig loads the config defined in `ProjectDir/.kepler/config.yaml`
+// and prepares the Dockerfile template defined in `ProjectDir/.kepler/Dockerfile`
+// On success, it will return a struct with all the required information
+// Otherwise, review the returned error message
+func CreateConfig(projectDir string) (*Config, error) {
+	// If we have been told to use the local directory,
+	// Set it to blank so that we don't mess up references
+	if projectDir == "." {
+		projectDir = ""
+	}
+	conf := path.Join(projectDir, ".kepler/config.yaml")
+	if _, err := os.Stat(conf); os.IsNotExist(err) {
+		return nil, fmt.Errorf("Unable to find %s", conf)
+	}
+	b, err := ioutil.ReadFile(conf)
+	if err != nil {
+		return nil, err
+	}
+	config := Config{
+		Application: path.Base(projectDir),
+	}
+	if err = yaml.Unmarshal(b, &config); err != nil {
+		return nil, err
+	}
+	// Enforce that resources are not resolved for this if
+	// config type isn't defined
+	template := path.Join(projectDir, ".kepler/Dockerfile.tmpl")
+	if _, err = os.Stat(template); os.IsNotExist(err) {
+		return nil, fmt.Errorf("Expected file %s missing", template)
+	}
+	b, err = ioutil.ReadFile(template)
+	if err != nil {
+		return nil, err
+	}
+	config.Template = b
+	return &config, nil
+}
+
+func (conf *Config) prepareTemplate() ([]byte, error) {
+	resolverType := strings.ToLower(conf.Type)
+	resolverFunc, exist := Resolvers[resolverType]
+	if exist {
+		addtional, err := resolverFunc(conf.Application)
+		if err != nil {
+			return nil, err
+		}
+		conf.Resources = append(conf.Resources, addtional...)
+	}
+	t := template.Must(template.New("Dockerfile").Parse(string(conf.Template)))
+	dockerfile := &bytes.Buffer{}
+	err := t.Execute(dockerfile, conf)
+	return dockerfile.Bytes(), err
+}
+
+// CreateStandaloneFile strips all the templating from the original template
+// without doing any resource resolution and returns a byte stream that is used as the dockerfile
+func (conf *Config) CreateStandaloneFile() ([]byte, error) {
+	// Have to modify the application name to be dot to
+	// ensure it copies the files in the current directory
+	if err := conf.validate(); err != nil {
+		return nil, err
+	}
+	name := conf.Application
```

Comment:
This is a bit gross

Reply:
It is but it is somewhat needed as when you create a Dockerfile to use inside a directory (that is external from the meta repo).

Given the case of the template looking like this:
```
FROM sratch

COPY {{.Application} /src/app
...
```
If I didn't enforce that the application name is `"."` then it would break.
I store the actual name so if any down streams calls are made, it has the name it expected.